### PR TITLE
Fix and improve chart interaction

### LINKED ISSUE #3278: Toggling chart interaction in multi-window plot fails
Slightly modified example from the original discussion thread to reproduce the issue.

```python3
import numpy as np
import pyvista as pv

shape = (4, 1)  # 4 by 1 grid
row_weights = [1, 1, 1, 1]
plotter = pv.Plotter(shape=shape, row_weights=row_weights)
rng = np.random.default_rng(1)

for i in range(4):
    plotter.subplot(i, 0)
    # Dummy
    x = rng.standard_normal(100)
    y = rng.standard_normal(100)
    chart = pv.Chart2D()
    chart.grid = True
    chart.background_color = "white"
    chart.scatter(x, y, size=10, style="o", color="cyan")
    plotter.add_chart(chart)

plotter.show()
            
```

Expected behaviour: all charts should become (non-)interactive after a double click.
Actual behaviour: only the bottom most chart correctly becomes (non-)interactive on a double click. The others remain in their initial state (non-interactive).

_Originally posted by @apaaris in https://github.com/pyvista/pyvista/discussions/3252#discussioncomment-3528435_

### PR DESCRIPTION
**Pull request by dcbr**
### Overview

Improve chart interaction and partially fixes a chart interaction bug.

### Details

- Fix chart interaction bug when there are multiple subplots, resolves #3278; but also requires an upstream (VTK) bug fix, see [vtk#18763](https://gitlab.kitware.com/vtk/vtk/-/issues/18763).
- Support interaction with multiple charts in the same renderer (this was an arbitrary constraint to simplify the initial implementation).
- Additionally, support enabling chart interaction from code, as requested recently in #3702.
- Added render event handling in the ``Renderer`` class.
- Fixed triple click being registered as a second double click.
- Added ``Chart`` type, to simplify doc string parameter types.

### TODO

- [x] Finish documentation.
- [x] Add separate chart border/background properties for when a chart is interactive, to make it visually easier to see which charts are interactive. I'll first finalize #2868 and then come back to this.

### CODE REVIEW

File: pyvista/plotting/render_window_interactor.py [hunk 1]
```diff
@@ -305,26 +310,66 @@ def update_style(self):
...
 
+        Parameters
+        ----------
+        scene : vtkContextScene, optional
+            The scene to interact with or ``None`` to stop interaction with any scene.
+
+        """
         # Set scene to interact with or reset it to stop interaction (otherwise crash)
+        if vtk_version_info < (9, 3, 0):
+            # TODO: determine in which VTK version !9811 will be available
```
Comment (dcbr) [inline pyvista/plotting/render_window_interactor.py:366]: @banesullivan Do you know whether the next VTK version will be 9.2.6 or 9.3.0? Then I can update this check accordingly :)

Comment (banesullivan) [reply pyvista/plotting/render_window_interactor.py:366]: I don't think that was backported to the `release` branch, so I'd put my money on `9.3`

File: pyvista/plotting/render_window_interactor.py [hunk 2]
```diff
@@ -305,26 +310,65 @@ def update_style(self):
             self._style_class = _style_factory(self._style)(self)
         self.interactor.SetInteractorStyle(self._style_class)
 
-    def _toggle_context_style(self, mouse_pos):
-        """Toggle the context style for chart interaction.
+    def _toggle_chart_interaction(self, mouse_pos):
+        """
+        Toggle interaction with indicated charts.
```
Comment (dcbr) [inline pyvista/plotting/render_window_interactor.py:315]:
```suggestion
        """Toggle interaction with indicated charts.
```

File: pyvista/plotting/renderer.py [hunk 3]
```diff
@@ -293,9 +294,52 @@ def _charts(self):
         # lazy instantiation here to avoid creating the charts object unless needed.
         if self.__charts is None:
             self.__charts = Charts(self)
-            self.AddObserver("StartEvent", partial(try_callback, self._render_event))
+            self.add_render_event(self._before_render_event, "before")
         return self.__charts
 
+    def add_render_event(self, callback, event="before"):
```
Comment (dcbr) [inline pyvista/plotting/renderer.py:300]: I added these methods at an earlier stage in this PR for a workaround that is no longer needed at this point. I think it's still ok to keep these in though?

**Review by dcbr — COMMENTED** [review]
This is ready for review now!

File: pyvista/plotting/render_window_interactor.py [hunk 4]
```diff
@@ -305,26 +310,65 @@ def update_style(self):
...
+        Set the context style interactor or switch back to previous interactor style.
 
+        Parameters
+        ----------
+        scene : vtkContextScene, optional
+            The scene to interact with or ``None`` to stop interaction with any scene.
+
+        """
         # Set scene to interact with or reset it to stop interaction (otherwise crash)
+        if vtk_version_info < (9, 3, 0):
```
Comment (dcbr) [inline pyvista/plotting/render_window_interactor.py:365]:
```suggestion
        if vtk_version_info < (9, 3, 0):  # pragma: no cover
```

File: pyvista/plotting/renderer.py (hunk 3 above)
Comment (banesullivan) [reply pyvista/plotting/renderer.py:300]: I think this method should be `interaction_event`, not `add_render_event` as the VTK event types are `StartEvent` and `EndEvent`. Further, this obfuscates the `add_on_render_callback()` I added in #3385.

Comment (banesullivan) [reply pyvista/plotting/renderer.py:300]: Oof, I think I may need to refactor `add_on_render_callback()` to be on the `Renderer`...

File: pyvista/plotting/renderer.py [hunk 5]
```diff
@@ -2939,6 +3002,7 @@ def remove_environment_texture(self):
     def close(self):
         """Close out widgets and sensitive elements."""
         self.RemoveAllObservers()
+        self._render_events.clear()
```
Comment (banesullivan) [inline pyvista/plotting/renderer.py:3005]: Thanks for pointing this out... we should do the same for `add_on_render_callback()`'s callbacks on close

**Review by banesullivan — COMMENTED** [review]
@dcbr, this is looking good. Some minor concerns about the render event listeners.

Do you want to land this today for `0.38`?

**Pull request merged**